rfd/0025-hsm.md

```diff
@@ -0,0 +1,378 @@
+}
+```
+
+This looks like it supports multiple private keys, but in practice only the
+first item of each list is used for signing. Since we can't add a `bool
+signing` flag to SSH keys (because they are byte slices), there isn't an
+incremental way to mark multiple signing keys.
+
+I propose revamping the schema entirely and doing a migration to move existing
+data:
+
+```proto
+message CertAuthoritySpecV2 {
+    // Note: some fields are omitted below.
+
+    // Deprecated fields for backwards-compatibility.
+    // Delete in teleport v8.
+    repeated bytes CheckingKeys = 3 [ (gogoproto.jsontag) = "checking_keys,omitempty" ];
+    repeated bytes SigningKeys = 4 [ (gogoproto.jsontag) = "signing_keys,omitempty" ];
+    repeated TLSKeyPair TLSKeyPairs = 7 [ (gogoproto.nullable) = false, (gogoproto.jsontag) = "tls_key_pairs,omitempty" ];
+    repeated JWTKeyPair JWTKeyPairs = 10 [ (gogoproto.nullable) = false, (gogoproto.jsontag) = "jwt_key_pairs,omitempty" ];
+
+    // New fields in teleport v7.
+    CAKeySet SigningKeys = 11;
+    CAKeySet TrustedKeys = 12;
+}
+
+message CAKeySet {
+    repeated SSHKeyPair SSH = 1;
+    repeated TLSKeyPair TLS = 2;
+    repeated JWTKeyPair JWT = 3;
+}
+
+message SSHKeyPair {
+    bytes PublicKey = 1;
+    bytes PrivateKey = 2;
+    PrivateKeyType PrivateKeyType = 3;
+}
+
+// TLSKeyPair already exists.
+message TLSKeyPair {
+    bytes Cert = 1;
+    bytes Key = 2;
+    // New field.
+    PrivateKeyType KeyType = 3;
+}
+
+// JWTKeyPair already exists.
+message JWTKeyPair {
+    bytes PublicKey = 1;
+    bytes PrivateKey = 2;
+    // New field.
+    PrivateKeyType PrivateKeyType = 3;
+}
+
+enum PrivateKeyType {
+    RAW = 0;
+    PKCS11 = 1;
+}
+```
+
+Clients will use a concatenation of both `CAKeySet`s. Auth servers will use
+`SigningKeys` to sign new certificates. CA rotation can now swap `SigningKeys`
+and `TrustedKeys` instead of relying on indexes.
+
+When `PrivateKeyType` is set to `PKCS11`, the `PrivateKey`/`Key` fields contain
+encoded information about the PKCS#11 key, as a JSON object prefixed with
+`pkcs11:`:
+
+```
+pkcs11:{"host_id": "host-uuid", "key_id": "key-uuid"}
+```
+
+Where `host-uuid` is the UUID of the Auth server that created the key and
+`key_id` is the UUID of the key set in the `CKA_ID` attribute on the HSM.
+
+Auth servers, when signing, will use:
+- if configured with PKCS#11, the `SigningKeys` PKCS#11 key that belongs to
+  their host (matched via `host_id` field)
+- if _not_ configured with PKCS#11, the first non-PKCS#11 item in `SigningKeys`
+
+### Configuration
+
+Auth servers need PKCS#11 configuration, passed in via `teleport.yaml`:
+
+```yaml
+auth_service:
+  ca_key_params:
+    pkcs11:
+      # Path to PKCS#11 module from the HSM vendor.
+      module_path: "/path/to/pkcs11.so"
+      # CKA_LABEL of the hSM, set during HSM initialization.
+      label: "myhsm"
+      # PIN for connecting to the HSM, set during HSM initialization.
+      pin: "password"
```

Comment:
This is probably more of an implementation detail, but we should also allow reading this value from a file or even from the console when Teleport starts up to avoid having the plaintext in the config.

Reply:
Agree with @webvictim